### PR TITLE
Intial webscripts sub-generator

### PR DESCRIPTION
**Pull request by binduwavell**
@vprince1 & @jrosler could you guys review this one too. There is quite a lot here and sadly no tests for the actual webscript subgenerator. But I'm hoping to get this into master soon as I think it provides a pretty nice basis for other sub generators. It provides a common library of prompt filters and a couple of base classes. One that sets up some default properties on the generator and provides a mechanism for specifying options and arguments right in the prompt data structure. The other one extends the first and wraps up selecting a source module to apply the scaffold into.

### CODE REVIEW

**Comment by vprince1** [comment]
Is there a reason why prompt-filters.js is not in a utils/ folder?

**Comment by binduwavell** [comment]
> Is there a reason why prompt-filters.js is not in a utils/ folder?

~~Nope, I'll move it there, thanks for the suggestion!~~

Oh wait, there isn't a utils folder, what did you have in mind?

**Comment by binduwavell** [comment]
> what did you have in mind?

Discussed in person. We agreed we'd try to create a generators/common folder and move all the stuff from app that is not specifically intended for the app generator into common to make it clearer it is a shared resource for all generators.

**Comment by jrosler** [comment]
I think we should say something like this

Current POM’s and context files in the project for building Alfresco will be updated as part of this process. Yeoman will display “conflict <filename>” and ask you if you want to update each file, or you can type “a” to say yes to updating all. Type “h” to get a list of all options.

**Comment by jrosler** [comment]
I also think it’d be a good idea to default to “a” if we can do that.

**Comment by binduwavell** [comment]
> I also think it’d be a good idea to default to “a” if we can do that.

I reviewed the following code from yeoman: https://github.com/yeoman/generator/blob/master/lib/util/conflicter.js

It does not appear that we can affect the default value when conflicts occur.

I'm still looking to see how the conflicter is called in case we can exclude certain files from going through the conflicter.

**Comment by binduwavell** [comment]
> I'm still looking to see how the conflicter is called in case we can exclude certain files from going through the conflicter.

As far as I can tell from reviewing this yeoman code: https://github.com/yeoman/generator/blob/995b95502604d6743c83b1d1a7074dce5b3727e9/lib/base.js

It appears we can use force to disable all conflict checking. But, there does not appear to be a way to disable conflict checking for certain files (other than the .yo-rc and .yo-rc-global files which are hard coded as being excluded.)

**Comment by binduwavell** [comment]
> Current POM’s and context files in the project for building Alfresco will be updated as part of this process. Yeoman will display “conflict <filename>” and ask you if you want to update each file, or you can type “a” to say yes to updating all. Type “h” to get a list of all options.

I added the following informational message:

> INFO:  This sub-generator will update existing POM's and context files. Yeoman will display "conflict <filename>" and ask you if you want to update each file. Type "h" when prompted to get details about your resolution choices.

File: test/webscript.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+'use strict';
+var path = require('path');
+var assert = require('yeoman-assert');
+var helpers = require('yeoman-generator').test;
+
+describe('generator-alfresco:webscript', function () {
```
Comment (vprince1) [inline test/webscript.js:6]: Since everything in this file is commented, is it necessary?

Comment (binduwavell) [inline test/webscript.js:6]: Yes in as much as it needs to be filled out! So keeping ING the file is my
shame and reminder that it needs to happen.

**Comment by vprince1** [comment]
I created a new amp and tried to create a repo webscript for this new amp. I got the below error. I am attaching the .yo-rc.json file for reference and also the interview with answers.
[mytest-yo-rc.json.txt](https://github.com/binduwavell/generator-alfresco/files/183509/mytest-yo-rc.json.txt)

[mytest-webscript-questions-answers.txt](https://github.com/binduwavell/generator-alfresco/files/183518/mytest-webscript-questions-answers.txt)

? Provide a | separated list of <response @type> values
INFO:  Generating asdasd.get.desc.xml in /Users/vijayprince/Work/projects/del/mytest/customizations/mytest-repo-amp/src/main/amp/config/alfresco/extension/templates/webscripts/sdf/asd
readline.js:925
            throw err;
            ^

ReferenceError: /Users/vijayprince/Work/projects/generator-alfresco/generators/webscript/templates/desc.xml.ejs:4
    2| function optionalAttribute(prop, attrName) {
    3|   if(prop) {

> > 4|     %> <%- attrName %>="<%- prop %>"<%
> >     5|   }
> >     6| }
> >     7| -%>

transactionBuffersize is not defined
    at eval (eval at <anonymous> (/Users/vijayprince/Work/projects/generator-alfresco/node_modules/ejs/lib/ejs.js:485:12), <anonymous>:67:26)
    at returnedFn (/Users/vijayprince/Work/projects/generator-alfresco/node_modules/ejs/lib/ejs.js:514:17)
    at Object.exports.render (/Users/vijayprince/Work/projects/generator-alfresco/node_modules/ejs/lib/ejs.js:316:37)
    at copy.process (/Users/vijayprince/Work/projects/generator-alfresco/node_modules/mem-fs-editor/lib/actions/copy-tpl.js:14:18)
    at applyProcessingFunc (/Users/vijayprince/Work/projects/generator-alfresco/node_modules/mem-fs-editor/lib/actions/copy.js:13:16)
    at EditionInterface.exports._copySingle (/Users/vijayprince/Work/projects/generator-alfresco/node_modules/mem-fs-editor/lib/actions/copy.js:51:16)
    at EditionInterface.exports.copy (/Users/vijayprince/Work/projects/generator-alfresco/node_modules/mem-fs-editor/lib/actions/copy.js:23:17)
    at EditionInterface.module.exports [as copyTpl](/Users/vijayprince/Work/projects/generator-alfresco/node_modules/mem-fs-editor/lib/actions/copy-tpl.js:12:8)
    at null.<anonymous> (/Users/vijayprince/Work/projects/generator-alfresco/generators/webscript/index.js:847:17)
    at Array.forEach (native)
    at null.<anonymous> (/Users/vijayprince/Work/projects/generator-alfresco/generators/webscript/index.js:843:21)
    at null.<anonymous> (/Users/vijayprince/Work/projects/generator-alfresco/generators/subgenerator.js:71:29)
    at null.<anonymous> (/Users/vijayprince/Work/projects/generator-alfresco/node_modules/yeoman-generator/lib/base.js:239:7)
    at PromptUI.onCompletion (/usr/local/lib/node_modules/yo/node_modules/inquirer/lib/ui/prompt.js:57:10)
    at AnonymousObserver.Rx.AnonymousObserver.AnonymousObserver.completed (/usr/local/lib/node_modules/yo/node_modules/rx-lite/rx.lite.js:1550:12)
    at AnonymousObserver.Rx.internals.AbstractObserver.AbstractObserver.onCompleted (/usr/local/lib/node_modules/yo/node_modules/rx-lite/rx.lite.js:1489:14)

**Comment by binduwavell** [comment]
Thanks @vprince1 I believe I have the issue resolved. Actually I saw two issues. Obviously the stack trace, but also the informational text for buffersize was displayed even though the question was not asked. Both of these should be resolved now.

**Comment by vprince1** [comment]
LGTM

**Comment by vprince1** [comment]
I created #30 to track the issue @jrosler mentioned about forcible overwriting files like pom.xml.

**Pull request merged**